perf/ring.go

```diff
@@ -158,3 +195,78 @@ func (rr *ringReader) Read(p []byte) (int, error) {
 
 	return n, nil
 }
+
+func (rr *ringReader) readOverwritable(p []byte) (int, error) {
+	read := rr.tail
+
+	// If size is 0, it means we read all the data avaiable in the buffer and jump
+	// on 0 data:
+	//
+	// prod_pos                         read_pos
+	//     |                                |
+	//     V                                V
+	// +---+------+----------+-------+------+
+	// |   |D....D|C........C|B.....B|A....A|
+	// +---+------+----------+-------+------+
+	if len(p) == 0 {
+		rr.tail = 0
+
+		return 0, io.EOF
+	}
+
+	// If adding the size to the current consumer position makes us wrap the
+	// buffer, it means we already did "one loop" around the buffer.
+	// So, the pointed data would not be usable:
+	//
+	//                               prod_pos
+	//                   read_pos----+   |
+	//                               |   |
+	//                               V   V
+	// +---+------+----------+-------+---+--+
+	// |..E|D....D|C........C|B.....B|A..|E.|
+	// +---+------+----------+-------+---+--+
+	if read-rr.head+uint64(len(p)) > rr.mask {
+		rr.tail = 0
+
+		return 0, io.EOF
+	}
+
+	size := uint32(len(p))
+	previousReadMasked := read & rr.mask
+	read += uint64(size)
+
+	// If adding the event size to the current
+	// consumer position makes us going from end of the buffer toward the
+	// start, we need to copy the rr.ring in two times:
+	// 1. First from previous_read_pos until end of the buffer.
+	// 2. Second from start of the buffer until read_pos.
+	//
+	// read_pos                  previous_read_pos
+	//     |                             |
+	//     V                             V
+	// +---+------+----------+-------+---+--+
+	// |..E|D....D|C........C|B.....B|A..|E.|
+	// +---+------+----------+-------+---+--+
+	// This code snippet was highly inspired by gobpf:
+	// https://github.com/iovisor/gobpf/blob/16120a1bf4d4abc1f9cf37fecfb86009a1631b9f/elf/perf.go#L148
+	if (read & rr.mask) < previousReadMasked {
```

Comment:
This is not necessary, since it's OK for read to return `n < len(p)` (see https://pkg.go.dev/io#Reader). It's the callers responsibility to call Read until the buffer is filled, which we do in `readRawSample`, etc via `io.ReadFull`: https://github.com/cilium/ebpf/blob/8c2039b96a96db0a60f9a897858c5f0fb70eb18f/perf/reader.go#L113-L115

Instead, copy what the current implementation does and truncate the read: https://github.com/cilium/ebpf/blob/8c2039b96a96db0a60f9a897858c5f0fb70eb18f/perf/ring.go#L142-L150